source/brailleDisplayDrivers/dotPad/driver.py

```diff
@@ -157,20 +193,76 @@ def _sendCommand(
 			return response.data
 		return b""
 
-	def _onReceive(self, header1: bytes):
-		if ord(header1) != DP_PacketSyncByte.SYNC1:
-			raise RuntimeError(f"Bad {header1=}")
-		header2 = self._dev.read(1)
-		if ord(header2) != DP_PacketSyncByte.SYNC2:
-			raise RuntimeError(f"bad {header2=}")
-		length = struct.unpack(">H", self._dev.read(2))[0]
-		packetBody = self._dev.read(length)
+	def _onReceive(self, data: bytes):
+		"""Handle received data from either Serial (1 byte) or BLE (full packets).
+
+		This method buffers incoming data and extracts complete packets as they arrive.
+		It works with both byte-at-a-time delivery (Serial) and packet-based delivery (BLE).
+
+		:param data: Received data (1 byte for Serial, variable length for BLE)
+		"""
+		self._receiveBuffer.extend(data)
+
+		# Safety check: prevent unbounded buffer growth from malformed streams
+		if len(self._receiveBuffer) > self.MAX_PACKET_SIZE:
+			log.warning(
+				f"Receive buffer exceeded {self.MAX_PACKET_SIZE} bytes, discarding data and resyncing",
+			)
+			self._receiveBuffer.clear()
+			return
+
+		# Extract and process all complete packets from the buffer
+		while len(self._receiveBuffer) >= 4:  # Minimum: SYNC1 + SYNC2 + length (2 bytes)
+			# Check for first sync byte
+			if self._receiveBuffer[0] != DP_PacketSyncByte.SYNC1:
+				# Discard bad byte and try to resynchronize
+				log.debug(f"Bad first sync byte: 0x{self._receiveBuffer[0]:02x}, discarding")
+				self._receiveBuffer.pop(0)
+				continue
+
+			# Check for second sync byte
+			if self._receiveBuffer[1] != DP_PacketSyncByte.SYNC2:
+				# Discard first byte and try again
+				log.debug(f"Bad second sync byte: 0x{self._receiveBuffer[1]:02x}, discarding")
+				self._receiveBuffer.pop(0)
+				continue
+
+			# Extract packet length from header
+			packetLength = struct.unpack(">H", bytes(self._receiveBuffer[2:4]))[0]
+			totalLength = 4 + packetLength  # header (4 bytes) + body
+
+			# Check if we have the complete packet
+			if len(self._receiveBuffer) < totalLength:
+				# Not enough data yet, wait for more
+				break
+
+			# Extract complete packet
+			packet = bytes(self._receiveBuffer[:totalLength])
+			self._receiveBuffer = self._receiveBuffer[totalLength:]
+
+			# Process the packet body (skip 4-byte header: SYNC1, SYNC2, length)
+			try:
+				self._processPacket(packet[4:])
+			except Exception:
+				log.error("Error processing packet", exc_info=True)
```

Comment:
This will re-enter the loop, right? Can't this create an infinite loop we'll never break from?